File: LeetCode/Array/Python/Leetcode42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l, r = [0] * n, [0] * n
        l[0], r[n - 1] = height[0], height[-1]
        for i in range(1, n):
            l[i] = max(l[i - 1], height[i])
        for i in range(n - 2, 0, -1):
            r[i] = max(r[i + 1], height[i])
        ans = 0
        for i in range(1, n - 1):
            ans += min(l[i], r[i]) - height[i]
        return ans

File: LeetCode/Array/Rust/Leetcode42-trapping-rain-water.rs
impl Solution {
  pub fn trap(height: Vec<i32>) -> i32 {
    let n = height.len();
    let (mut l, mut r) = (vec![0; n], vec![0; n]);
    l[0] = height[0];
    r[n - 1] = height[n - 1];
    for i in 1..n {
      l[i] = l[i - 1].max(height[i]);
    }
    for i in (0..n - 1).rev() {
      r[i] = r[i + 1].max(height[i]);
    }
    let mut ans = 0;
    for i in 1..n - 1 {
      ans += l[i].min(r[i]) - height[i];
    }
    ans
  }
}
